net: lwm2m: reset obj_inst/res_inst data structure when delete

rbtchc opened the pull request:
When a request demands to create a new object instance, it will search whether the request object instance exists or not. However, the current implementation does not reset the lwm2m_engine_obj_inst at the time it is deleted. It only removes the object instance from the sys list.
    
Correct the behavior by resetting both object instance and resource instances at the time it's deleted. Also, consolidate function lwm2m_delete_handler() and lwm2m_delete_obj_inst().
    
To reproduce the issue, try to create light control object instance (/3301), delete the created instance and create it again. You shall find following error message dumped.
> [ipso_light_control] [ERR] light_control_create: Can not create instance - already existing: 0
> [lib/lwm2m_engine] [ERR] lwm2m_create_obj_inst: unable to create obj - 3311 instance 0

rbtchc committed:
net: lwm2m: reset obj_inst/res_inst data structure when delete

When a request demands to create a new object instance, it will search
whether the request object instance exists or not. However, current
implementation does not reset the lwm2m_engine_obj_inst at the time it
is deleted. It only removes the object instance from the sys list.

Correct the behavior by resetting both object instance and resource
instances at the time it's deleted. Also, consolidate function
lwm2m_delete_handler() and lwm2m_delete_obj_inst().

To reproduce the issue, try to create light control object instance
(/3301), delete the created instance and create it again. You shall find
following error message dumped.
> [ipso_light_control] [ERR] light_control_create: Can not create
  instance - already existing: 0
> [lib/lwm2m_engine] [ERR] lwm2m_create_obj_inst: unable to create obj -
  3311 instance 0

Signed-off-by: Robert Chou <robert.ch.chou@acer.com>

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -430,10 +431,21 @@ int lwm2m_delete_obj_inst(u16_t obj_id, u16_t obj_inst_id)
 	engine_unregister_obj_inst(obj_inst);
 	obj->instance_count--;
+#ifdef CONFIG_LWM2M_RD_CLIENT_SUPPORT
+	engine_trigger_update();
```
I realize that this isn't a change you made, but we probably shouldn't trigger the registration update until after we've called the obj->delete_cb and reset the resources.

rbtchc commented:
Revised to trigger update at the end of the call.
Diff: https://hastebin.com/qehasayomo.cs

rbtchc committed:
net: lwm2m: fix obj/obj_inst/observer sys_slist_t corrupted when remove

We were using sys_slist_remove() to remove object, object instance and
observer w/o passing the previous sys_snode_t to it (NULL).
This will instruct the function to treat the node as the list head and
result in unexpected behavior after the removal.

Correct it by using sys_slist_find_and_remove() or passing the previous
node to the function.

Signed-off-by: Robert Chou <robert.ch.chou@acer.com>

rbtchc commented:
Add one more patch to fix incorrectly removing object/object instance/observer_node from the sys_slist_t

rbtchc committed:
net: lwm2m: clean up observer when object/object instance is removed

We should stop sending out notification to the peer when the
object/object instances requested to be observed is removed

Signed-off-by: Robert Chou <robert.ch.chou@acer.com>

rbtchc committed:
net: lwm2m: return 4.04 (not found) when obj/obj_inst/res not exist

We did not check the requested object/object instance/resource exists or
not before we adding an observer. Correct it by checking the existence
first.

Signed-off-by: Robert Chou <robert.ch.chou@acer.com>

rbtchc committed:
net: lwm2m: propogate errors to the caller when using TLV writer

Errors has been ignored when using TLV writer to create/write object
instance/resources. Modify to propagate the error back to the caller.

To reproduce the issue, try to create IPSO light control object
instances twice. Since the default instance count is 1, the second one
should be rejected and responded w/ error. But the current
implementation will respond w/ 2.04.

Signed-off-by: Robert Chou <robert.ch.chou@acer.com>

rbtchc commented:
Turns out I found several other issues while testing the delete / create object instances, observe/cancel observe object instance/resources. Three more patches are submitted for review.

rbtchc commented:
@mike-scott these four commits should be independent from other PRs. If possible, please take a look and see if we can merge these into the trunk. These are mainly focusing on correcting the behavior of create / delete object instances, observe/unobserver object instances and resources.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -1674,32 +1686,17 @@ static int lwm2m_exec_handler(struct lwm2m_engine_obj *obj,
-		return obj->delete_cb(path->obj_inst_id);
-	}
+	path = context->path;
+	SYS_LOG_DBG(">> DELETE [path:%u/%u/%u(%u)]",
```
I left this in the original code by accident.  Can you remove it here?

rbtchc replied:
sure

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -333,6 +347,22 @@ static void engine_register_obj_inst(struct lwm2m_engine_obj_inst *obj_inst)
+	/* remove observer instances accordingly */
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(
+			&engine_observer_list, obs, tmp, node) {
+		if (obs->path.obj_id != obj_inst->obj->obj_id ||
```
This logic would be clearer written as:
(!(obs->path.obj_id == obj_inst->obj->obj_id && obs->path.obj_inst_id == obj_inst->obj_inst_id))

mike-scott replied:
A small helper to remove observers from the list would eliminate code duplication here.  It could take params: obj_id and obj_inst_id.  The obj_inst_id could be s32_t and accept -1 as an option meaning to skip checking it (match only obj_id only).

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -195,6 +200,39 @@ static int engine_add_observer(struct net_context *net_ctx,
+		return -ENOENT;
+	}
+
+	/* check if object instane exists */
```
s/instane/instance

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -195,6 +200,39 @@ static int engine_add_observer(struct net_context *net_ctx,
 
+	/* check if object exists */
+	if (!get_engine_obj(path->obj_id)) {
+		SYS_LOG_DBG("unable to find obj: %u", path->obj_id);
```
SYS_LOG_ERR

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -195,6 +200,39 @@ static int engine_add_observer(struct net_context *net_ctx,
+		obj_inst = get_engine_obj_inst(path->obj_id,
+					       path->obj_inst_id);
+		if (!obj_inst) {
+			SYS_LOG_DBG("unable to find obj_inst: %u/%u",
```
SYS_LOG_ERR

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -195,6 +200,39 @@ static int engine_add_observer(struct net_context *net_ctx,
+		}
+
+		if (i == obj_inst->resource_count) {
+			SYS_LOG_DBG("unable to find res_id: %u/%u/%u",
```
SYS_LOG_ERR

mike-scott reviewed (commented):
Nice overall patchset.  Minor comments.

rbtchc commented:
Update according to @mike-scott 's review.
Here are the changes: https://hastebin.com/umuvucepop.php

mike-scott reviewed (approved):
Thanks @rbtchc LGTM
